lib/fh_cluster.js

```diff
@@ -62,12 +64,163 @@ module.exports = function fhCluster(workerFunc, optionalNumWorkers, optionalBack
+    log.info('Worker #', worker.id, 'exited. Will retry in',
+      nextRetry, 'ms');
+
+    setTimeout(function() {
+      cluster.fork();
+      resetBackoffResetTimeout(strategy, backoffResetTimeout);
+    }, nextRetry);
+  };
+}
+
+function setOnWorkerListeningHandler(worker, address) {
+  var host = address.address || address.addressType === 4 ? '0.0.0.0' : '::';
+  var addr = host + ':' + address.port;
+  log.info('Cluster worker',  worker.id, 'is now listening at', addr);
+
+  //A new process is listening, need to check if any bound tasks need to be assigned.
+  boundTasks.setUpBoundTasks(singleWorkerTasks, cluster);
+}
+
+function setupBoundsTasks(){
+  //If a worker exits, check that all bound workers are on a running worker.
+  boundTasks.setUpBoundTasks(singleWorkerTasks, cluster);
+}
+
+function resetBackoffResetTimeout(backoffStrategy, timeout) {
+  if (timeout) {
+    clearTimeout(timeout);
+  }
+
+  return setTimeout(function() {
+    backoffStrategy.reset();
+  }, 60*60*1000);
+}
+
+function cleanMasterShutdown(type) {
+  return function (cb) {
+    log.debug('exiting master with ' + type + ' [' + process.pid + ']');
+    eachWorker(function(worker) {
+      worker.send('shutdown');
+    });
+    waitForAllDead(function() {
+      log.debug('exited master with ' + type + ' [' + process.pid + ']');
+      if (cb) return cb();
+      process.exit(0);
+    });
+  }
+}
+
+function waitForAllDead(cb) {
+  var someAlive = false;
+  eachWorker(function(worker) {
+    if (!isWorkerDead(worker)) {
+      log.debug('worker ' + worker.id + ' not dead');
+      someAlive = true;
+      setTimeout(function() {
+        waitForAllDead(cb);
+      }, 50);
+    }
+  });
+  if (!someAlive) {
+    log.debug('all workers dead');
+    cb();
+  }
+}
+
+function isWorkerDead(worker ) {
+  return ('isDead' in worker) ? worker.isDead() : worker.state === 'dead';
+}
+
+function eachWorker(callback) {
+  for (var id in cluster.workers) {
+    callback(cluster.workers[id]);
+  }
+}
+
+/**********************************************/
+/* WORKER - code that powers worker processes */
+/**********************************************/
+
+function startWorker(workerFunc, gracefullShutdownFn) {
+  cluster.worker.on('message', function(message){
+    if(_.isString(message.startEventId)){
+      //Need to do something
+      var task = _.findWhere(singleWorkerTasks, {startEventId: message.startEventId});
+      if(!task){
+        return;
+      }
+
+      //Execute the required task.
+      task.workerFunction(cluster.worker);
+    }
+  });
+  process.on('SIGTERM', cleanWorkerShutdown('SIGTERM'));
+  process.on('SIGHUP', cleanWorkerShutdown('SIGHUP'));
+  process.on('SIGINT', cleanWorkerShutdown('SIGINT'));
+  process.on('message', function(msg) {
+    if (msg === 'shutdown') {
+      cleanWorkerShutdown('shutdown')(gracefullShutdownFn);
+    }
+  });
+
+  workerFunc(cluster.worker);
+}
+
+function cleanWorkerShutdown(type) {
+  return function (cb) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Add a semicolon at the end of this statement. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ASemicolon)